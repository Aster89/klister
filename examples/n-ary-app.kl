#lang kernel

[import [shift (only kernel cons-list-syntax lambda pure quote syntax-case list-syntax export import) 1]]



[define-macros
  ([my-app
    [lambda [stx]
      (syntax-case stx
        [[cons _ fun-and-args]
         (syntax-case fun-and-args
           [[cons fun args]
            (syntax-case args
              [() [pure fun]]
              [[cons arg0 more-args]
               [pure
                [cons-list-syntax
                  [quote my-app]
                  [cons-list-syntax
                    [list-syntax [[quote #%app] fun arg0] stx]
                    more-args
                    stx]
                  stx]]])])])]]
   [my-lam
    [lambda [stx]
     (syntax-case stx
       [[list [_ args body]]
        (syntax-case args
          [() [pure body]]
          [[cons arg0 more-args]
           [pure [list-syntax [[quote lambda]
                               [list-syntax [arg0] arg0]
                               [list-syntax [[quote my-lam] more-args body] stx]]
                           stx]]]
          [[list [arg]]
           [pure [list-syntax [[quote lambda] args body] stx]]])])]])]

[define-macros
  ([#%app
    [lambda [stx]
     (syntax-case stx
      [[cons _ args]
       [pure [cons-list-syntax [quote my-app] args stx]]])]])]

[define-macros
  ([lambda
    [lambda [stx]
     (syntax-case stx
      [[list [_ args body]]
       [pure [list-syntax [[quote my-lam] args body] stx]]])]])]

[define id [lambda (x) x]]

[define const [lambda (x y) x]]

[example (const [quote a] [quote b])]

[define compose [lambda (f g x) (f (g x))]]

[example (compose id)]
[example (compose id id)]
[example (compose id id id)]

[meta [export import]]
[export #%module]
[export #%app]
[export const]
[export lambda]
[export define]
[export example]
[export define-macros]
[export quote]
[export meta]
[export import]
[export >>=]
[export pure]
[export if]
[export syntax-error]
[export syntax-case]
[export list-syntax]
[export cons-list-syntax]
[export empty-list-syntax]
[export free-identifier=?]
[export log]
[export export]
