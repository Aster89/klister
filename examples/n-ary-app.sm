#lang kernel

[define-macros
  ([my-app
    [lambda [stx]
      (syntax-case stx
        [[cons _ fun-and-args]
         (syntax-case fun-and-args
           [[cons fun args]
            (syntax-case args
              [() [pure fun]]
              [[cons arg0 more-args]
               [pure
                [cons-list-syntax
                  [quote my-app]
                  [cons-list-syntax
                    [vec-syntax [[quote #%app] fun arg0] stx]
                    more-args
                    stx]
                  stx]]])])])]])]

[define-macros
  ([#%app
    [lambda [stx]
     (syntax-case stx
      [[cons _ args]
       [pure [cons-list-syntax [quote my-app] args stx]]])]])]


[define const [lambda [x] [lambda [y] x]]]

[example (const [quote a] [quote b])]
