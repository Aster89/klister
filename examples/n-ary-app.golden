#[n-ary-app.kl:90.18-90.19]<a> : Syntax
#<closure> : ∀α β. ((α → β) → (α → β))
#<closure> : ∀α. (α → α)
#<closure> : ∀α. (α → α)
#<closure> : ∀α β. ((α → β) → (α → β))
#<closure> : ∀α. (α → α)
#<closure> : ∀α. (α → α)
